cohortextractor/mssql_utils.py

```diff
@@ -88,3 +89,25 @@ def mssql_sqlalchemy_engine_from_url(url):
     params["drivername"] = "mssql+pyodbc"
     params["query"] = {"driver": "ODBC Driver 17 for SQL Server"}
     return sqlalchemy.create_engine(URL(**params))
+
+
+def dbapi_cursor_to_csv_file(cursor, filename, batch_size=None, row_callback=None):
+    """
+    Takes a DB-API comptabile cursor object with a result set and writes those
+    results to a CSV file, calling `row_callback` (if defined) on each row as
+    it does so
+    """
+    if batch_size is None:
+        batch_size = cursor.arraysize
+    if row_callback is None:
+        row_callback = lambda x: None
+    with open(filename, "w", newline="") as csvfile:
+        writer = csv.writer(csvfile)
+        writer.writerow([x[0] for x in cursor.description])
+        while True:
+            result_batch = cursor.fetchmany(batch_size)
+            if not result_batch:
+                break
+            for row in result_batch:
+                row_callback(row)
+                writer.writerow(row)
```

Comment:
Do we want these two lines the other way round?  If the callback fails, it'd be nice to know which line it failed on.

Reply:
Ah, good thinking!